src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,189 @@
+            this.last = node0;
+            this.size++;
+            return;
+        }
+        Node<T> node = new Node<>(last, null, value);
+        this.last.nextNode = node;
+        this.last = node;
+        this.size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == 0 && this.size == 0) {
+            this.add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, this.first.nextNode, value);
+            this.first.prevNode = newNode;
+            this.first = newNode;
+            this.size++;
+            return;
+        }
+        this.validIndex(index - 1);
+        if (index != this.size) {
+            Node<T> node = this.indexedNode(index);
+            Node<T> newNode = new Node<>(node.prevNode, node.nextNode, value);
+            node.prevNode = newNode;
+            node = newNode.prevNode;
+            node.nextNode = newNode;
+            this.size++;
+            return;
+        }
+        this.add(value);
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (T i : list) {
+            this.add(i);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        this.validIndex(index);
+        return this.indexedNode(index).value;
     }
 
     @Override
     public void set(T value, int index) {
-
+        this.validIndex(index);
+        if (index == 0) {
+            this.first.value = value;
+            return;
+        }
+        if (index == this.size - 1) {
+            this.last.value = value;
+            return;
+        }
+        this.indexedNode(index).value = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        this.validIndex(index);
+        if (index == 0 && this.size == 1) {
+            Node<T> node = this.indexedNode(index);
+            this.last = null;
+            this.first = null;
+            T value = node.value;
+            this.size--;
+            return value;
+        }
+        if (index == 0) {
+            Node<T> nextNode = this.indexedNode(index).nextNode;
+            Node<T> node = this.indexedNode(index);
+            nextNode.prevNode = null;
+            this.first = nextNode;
+            T value = node.value;
+            this.size--;
+            return value;
+        }
+        if (index == this.size - 1) {
+            Node<T> prevNode = this.indexedNode(index).prevNode;
+            Node<T> node = this.indexedNode(index);
+            prevNode.nextNode = null;
+            this.last = prevNode;
+            T value = node.value;
+            this.size--;
+            return value;
+        }
+        Node<T> prevNode = this.indexedNode(index).prevNode;
+        Node<T> nextNode = this.indexedNode(index).nextNode;
+        Node<T> node = this.indexedNode(index);
+        prevNode.nextNode = nextNode;
+        nextNode.prevNode = prevNode;
+        T value = node.value;
```

Comment:
how about removing the link on previous and next node?

Reply:
That is unnecessary. We deleted links on this Node, Garbage Collector will delete it. 
Bogdan talked about it in a lecture.)